presto-main/src/main/java/io/prestosql/server/DynamicFilterService.java

```diff
@@ -0,0 +1,244 @@
+
+        return () -> {
+            AtomicReference<TupleDomain<ColumnHandle>> dynamicPredicates = new AtomicReference<>(TupleDomain.all());
+            for (DynamicFilters.Descriptor dynamicFilter : dynamicFilters) {
+                String filterId = dynamicFilter.getId();
+                getSummary(queryId, filterId)
+                        .map(summary -> translateSummaryToTupleDomain(filterId, summary, sourceColumnHandles))
+                        // Intersect dynamic filters' predicates when they become ready
+                        .ifPresent(tupleDomain -> dynamicPredicates.updateAndGet(
+                                constraint -> constraint.intersect(tupleDomain)));
+            }
+            return dynamicPredicates.get();
+        };
+    }
+
+    @Immutable
+    private static class SourceDescriptor
+    {
+        private final QueryId queryId;
+        private final String filterId;
+
+        public static SourceDescriptor of(QueryId queryId, String filterId)
+        {
+            return new SourceDescriptor(queryId, filterId);
+        }
+
+        private SourceDescriptor(QueryId queryId, String filterId)
+        {
+            this.queryId = requireNonNull(queryId, "queryId is null");
+            this.filterId = requireNonNull(filterId, "filterId is null");
+        }
+
+        public QueryId getQueryId()
+        {
+            return queryId;
+        }
+
+        public String getFilterId()
+        {
+            return filterId;
+        }
+
+        @Override
+        public boolean equals(Object other)
+        {
+            if (other == this) {
+                return true;
+            }
+            if (other == null || getClass() != other.getClass()) {
+                return false;
+            }
+
+            SourceDescriptor sourceDescriptor = (SourceDescriptor) other;
+
+            return Objects.equals(queryId, sourceDescriptor.queryId) &&
+                    Objects.equals(filterId, sourceDescriptor.filterId);
+        }
+
+        @Override
+        public int hashCode()
+        {
+            return Objects.hash(queryId, filterId);
+        }
+
+        @Override
+        public String toString()
+        {
+            return toStringHelper(this)
+                    .addValue(queryId)
+                    .addValue(filterId)
+                    .toString();
+        }
+    }
+
+    private static TupleDomain<ColumnHandle> translateSummaryToTupleDomain(String filterId, Domain summary, Map<String, ColumnHandle> sourceColumnHandles)
+    {
+        if (summary.isNone()) {
+            return TupleDomain.none();
+        }
+        return TupleDomain.withColumnDomains(ImmutableMap.<ColumnHandle, Domain>builder()
+                .put(sourceColumnHandles.get(filterId), summary)
+                .build());
+    }
+
+    private static Map<String, Symbol> extractSourceExpressionSymbols(List<DynamicFilters.Descriptor> dynamicFilters)
```

Comment:
make this return `ColumnHandle` directly

Reply:
done